modules/core/04-channel/keeper/upgrade_test.go

```diff
@@ -124,4 +128,180 @@ func (suite *KeeperTestSuite) TestChanUpgradeInit() {
 	}
 }
 
+// UpgradeInit will be run on chainA and startFlushUpgradeHandshake
+// will be called on chainB
+func (suite *KeeperTestSuite) TestStartFlushUpgradeHandshake() {
+	var (
+		path                *ibctesting.Path
+		upgrade             types.Upgrade
+		counterpartyChannel types.Channel
+		counterpartyUpgrade types.Upgrade
+	)
+
+	testCases := []struct {
+		name     string
+		malleate func()
+		// TODO: currently the code asserts based on an expected sdk error
+		// Allow support for asserting either the sdk error type or that the error is an upgrade error
+		expError error
+	}{
+		{
+			"success",
+			func() {},
+			nil,
+		},
+		{
+			"channel not found",
+			func() {
+				path.EndpointB.ChannelID = "invalid-channel"
+				path.EndpointB.ChannelConfig.PortID = "invalid-port"
+			},
+			types.ErrChannelNotFound,
+		},
+		{
+			"connection not found",
+			func() {
+				channel := path.EndpointB.GetChannel()
+				channel.ConnectionHops[0] = "invalid-connection"
+				path.EndpointB.SetChannel(channel)
+			},
+			connectiontypes.ErrConnectionNotFound,
+		},
+		{
+			"connection state is not in OPEN state",
+			func() {
+				conn := path.EndpointB.GetConnection()
+				conn.State = connectiontypes.INIT
+				path.EndpointB.SetConnection(conn)
+			},
+			connectiontypes.ErrInvalidConnectionState,
+		},
+		{
+			"failed verification for counterparty channel state due to incorrectly constructed counterparty channel",
+			func() {
+				counterpartyChannel.State = types.CLOSED
+			},
+			commitmenttypes.ErrInvalidProof,
+		},
+		{
+			"failed verification for counterparty upgrade due to incorrectly constructed counterparty upgrade",
+			func() {
+				counterpartyUpgrade.LatestSequenceSend = 100
+			},
+			commitmenttypes.ErrInvalidProof,
+		},
+		{
+			"upgrade sequence mismatch, endpointB channel upgrade sequence is ahead",
+			func() {
+				channel := path.EndpointB.GetChannel()
+				channel.UpgradeSequence++
+				path.EndpointB.SetChannel(channel)
+			},
+			types.ErrIncompatibleCounterpartyUpgrade,
```

Comment:
I'm not sure yet if there's a way we could return an `NewUpgradeError` here with the `ErrIncompatibleCounterpartyUpgrade` embedded and then do a check that the returned error is an upgrade error and contains the same parent err